django/contrib/gis/db/models/aggregates.py

```diff
@@ -49,8 +49,8 @@ def resolve_expression(
         self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False
     ):
         c = super().resolve_expression(query, allow_joins, reuse, summarize, for_save)
-        for expr in c.get_source_expressions():
-            if not hasattr(expr.field, "geom_type"):
+        for field in c.get_source_fields():
+            if not hasattr(field, "geom_type"):
```

Comment:
Unfortunately, it's not enough to make `filter` works on Oracle, as [`tolerance`](https://github.com/django/django/blob/06881341d48dd17a4fdf25afc96991de4fb3b1ac/django/contrib/gis/db/models/aggregates.py#L34-L45) sneaks into `filter`:
```
======================================================================
ERROR: test_union_filter (gis_tests.relatedapp.tests.RelatedGeoModelTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/django/django/test/testcases.py", line 1428, in skip_wrapper
    return test_func(*args, **kwargs)
  File "/django/tests/gis_tests/relatedapp/tests.py", line 407, in test_union_filter
    city = qs.get(name="Aurora")
  File "/django/django/db/models/query.py", line 625, in get
    num = len(clone)
  File "/django/django/db/models/query.py", line 379, in __len__
    self._fetch_all()
  File "/django/django/db/models/query.py", line 1892, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/django/django/db/models/query.py", line 90, in __iter__
    results = compiler.execute_sql(
  File "/django/django/db/models/sql/compiler.py", line 1549, in execute_sql
    sql, params = self.as_sql()
  File "/django/django/db/models/sql/compiler.py", line 736, in as_sql
    extra_select, order_by, group_by = self.pre_sql_setup(
  File "/django/django/db/models/sql/compiler.py", line 84, in pre_sql_setup
    self.setup_query(with_col_aliases=with_col_aliases)
  File "/django/django/db/models/sql/compiler.py", line 73, in setup_query
    self.select, self.klass_info, self.annotation_col_map = self.get_select(
  File "/django/django/db/models/sql/compiler.py", line 296, in get_select
    sql, params = self.compile(col)
  File "/django/django/db/models/sql/compiler.py", line 544, in compile
    sql, params = vendor_impl(self, self.connection)
  File "/django/django/contrib/gis/db/models/aggregates.py", line 43, in as_oracle
    return clone.as_sql(
  File "/django/django/contrib/gis/db/models/aggregates.py", line 25, in as_sql
    return super().as_sql(
  File "/django/django/db/models/aggregates.py", line 138, in as_sql
    condition = When(self.filter, then=source_expressions[0])
  File "/django/django/db/models/expressions.py", line 1365, in __init__
    raise TypeError(
TypeError: When() supports a Q object, a boolean expression, or lookups as a condition.
```

Reply:
> Unfortunately I don't have a Oracle on hand to test and there doesn't seem to be one in CI either.

There is an Oracle run in CI. It's called on demand, you need to leave a comment _"buildbot, test on oracle."_, see [CI wiki](https://code.djangoproject.com/wiki/CI).